Tower job template update

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Conversion of tower_job_template module off tower-cli

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collection

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 9.3.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Changing tower_group and tower_api for a patterned associate from generator

john-westcott-iv committed:
Compensating for default of '' for a JSON typed field

john-westcott-iv committed:
Adding associations to generator

john-westcott-iv committed:
Multiple changes to support tower_job_template

Created method for extracting errors from API return (extract_errors_from_response)

Added fields to JSON output to suplment changed field (i.e. updated, created, field_changes)

Added additional posts option (for surveys)

Extracted duplicate association logic into a method

john-westcott-iv committed:
Initial converstion of tower_job_template

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -94,6 +94,17 @@ def __init__(self, argument_spec, **kwargs):
+            if '__all__' in response['json']:
+                return response['json']['__all__'][0]
+            elif 'error' in response['json']:
+                return response['json']['error']
```
what about "detail" and "msg", I think those should be inspected too because it's sometimes a thing.

Also, best to do this in a loop.

And go ahead with the duck typing, if it's a list, then make it into a string. I don't like `[0]`, I'd rather do `\n'.join()`. Your choice for the separator.

john-westcott-iv replied:
Do you have examples of detail and msg? Are those at the root of the object returned? I'll take care of the loop and joins in the mean time.

john-westcott-iv committed:
Adding labels to Tower job template

john-westcott-iv committed:
Checking for __all__ and error in a loop

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
+
+                with open('/tmp/john', 'w') as f:
+                    import json
+                    f.write(json.dumps(post['data']))
```
Not going to lie, when I'm in these situations I `raise Exception(data)` and run the test

john-westcott-iv replied:
Yea, I should probably do something better but exceptions in Ansible can be a tad ugly to go through. Especially just when I want to try and dump a response data out. Removed the debug.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
+                    f.write(json.dumps(post['data']))
+
+                # If this item no longer has data, lets remove it
+                if post['data'] in ('', [], {}):
```
This could cause weird changed=True cases, since `existing_data['json'] == post['data']` will be unaware of cases which are different in syntax, but both indicate various version of nothing.

john-westcott-iv replied:
Do you have examples of that? This was a first crack at defining some kind of pattern, if we can't make it work we can try something else.

AlanCoding replied:
So I guess I'm not following this at all. This `if` should be written as `if not post['data']` unless I'm missing something. If `if 'data' in post and (not post['data'])`. Unless the intent is that `None` should not result in a DELETE. This is why the contract gets really hard to manage. There's a lot of code here for survey spec handling, I don't think it generalizes well, it doesn't align with WFJT surveys, and I would go back to the drawing board for everything with the spec.

john-westcott-iv replied:
Correct, None would represent that the survey_spec (or any other field) was just not provided to the module. So we should not delete the survey on a None condition. The module itself is currently catching survey_spec (or any other field) being None and choosing to not add the data to the additional post parameter. However we could also go the route of just passing None back into this method (especially if we are doing auto-generation).  So instead of blocks like this:
```
   additional_posts = []
    if survey_spec is not None:
        additional_posts.append({
            'endpoint_reference': 'survey_spec',
            'data': survey_spec,
        })
```
Followed with:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=additional_posts,
        )
```
We could just do something like:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=[
              {'endpoint_reference': 'survey_spec','data': survey_spec,}
            ],
        )
```

It looks like survey_spec is the same for WJT as it is for JT (unless of course I'm missing something) its even posted through the same survey_spec end point. I did a quick test converting WFJT module to use additional_posts and it seemed to work (granted it wasn't a very large survey).

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
+                    # Tower job template requires a 200
+                    # Left as an array as for future additional delete returns
+                    if response['status_code'] in [200]:
+                        self.json_output['deleted_additional_post'] = True
```
hmm... maybe it would be good to report _which_ endpoint was deleted, considering you wrote this to handle multiples

john-westcott-iv replied:
With the next push it will now spit out something like:
```
        "added_additional_post": "/api/v2/job_templates/112/survey_spec/",
        "added_an_association": [
            "/api/v2/job_templates/112/credentials/[1]",
            "/api/v2/job_templates/112/credentials/[14]",
            "/api/v2/job_templates/112/labels/[120]",
            "/api/v2/job_templates/112/labels/[121]"
        ],
```
I added the [<id>] to the "endpoint" so you would know better what was going on rather than just getting a bunch of meaningless /labels/ endpoints.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
+                if item.get('related', {}).get(association_type, None):
+                    self.modify_associations(item['related'][association_type], associations[association_type])
+                else:
+                    self.fail_json(msg="Association type of {0} is not valid for {1}".format(association_type, item_type))
```
this stuff should get cleaned up after rebasing

john-westcott-iv replied:
I feel like we have a little bit of a mess on our hands with all of the things flying around @beeankha and I are going to try and cleanup and rebase on Monday.

AlanCoding replied:
The existing item, `item` in this case will be `None` in some cases. This comes from the case where the object is created, and has associations, and the fix I applied to that doesn't look like it will work here. Some test coverage for this was introduced in https://github.com/ansible/awx/pull/6369. What changed with the processing of new items in this PR, and how is it expected that the case of create-and-associate will work with the call patterns as they were changed here?

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,532 @@
+                'credentials': credentials_ids,
+                'labels': labels_ids,
+            },
+            additional_posts=additional_posts,
```
To get a sense of how this compares to the current state, I wrote this, partially influenced by your code:

https://github.com/ansible/awx/pull/6291/commits/581d495f66ee831ffbf4f61422664a5e139c8886

I'm trying to grasp how your PR winds up with much more code, and it may be largely in the error handling.

john-westcott-iv committed:
Removing an open /tmp/john debug

AlanCoding commented:
I don't know if this is going to work for approval templates. These are created in a related POST from the WFJT node. This should be similar to what you're doing here, but the details go a little crazy.

Looking at what the UI does, it makes a POST:

url `/api/v2/workflow_job_template_nodes/10/create_approval_template/`

```json
{
  "name": "fffff",
  "description": null,
  "timeout": 0
}
```

In response, the API gives a 201 and a full serialization of the related object. Like:

```json
{
  "id": 17,
  "type": "workflow_approval_template",
  "url": "/api/v2/workflow_approval_templates/17/",
  "related": {
    "created_by": "/api/v2/users/1/",
    "modified_by": "/api/v2/users/1/",
    "jobs": "/api/v2/workflow_approval_templates/17/approvals/"
  },
  "summary_fields": {
    "workflow_job_template": {
      "id": 16,
      "name": "has approval node",
      "description": ""
    },
    "created_by": {
      "id": 1,
      "username": "admin",
      "first_name": "",
      "last_name": ""
    },
    "modified_by": {
      "id": 1,
      "username": "admin",
      "first_name": "",
      "last_name": ""
    }
  },
  "created": "2020-03-20T15:18:28.757683Z",
  "modified": "2020-03-20T15:18:28.757696Z",
  "name": "fffff",
  "description": "",
  "last_job_run": null,
  "last_job_failed": false,
  "next_job_run": null,
  "status": "never updated",
  "timeout": 0
}
```

@beeankha do you have any thoughts on how to handle approval nodes?

john-westcott-iv committed:
Added append_json_output method and put into use from removed_an_association, added_an_association, deleted_additional_post and added_additional_post

Rebase and drop redundant commits

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -518,27 +587,21 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
-        if associations is not None:
-            for association_type in associations:
-                self.modify_associations(response, associations[association_type])
+        self.process_associations(response['json'], item_type, associations)
```
This still isn't going to work if `existing_item` is non-null when the method is called.

https://github.com/ansible/awx/pull/6369

I'm sorry this is going to mess up the diff management even more. I really want us to focus on having these stable and testable as more features are added. If you want to put up a minor patch in a way that better aligns with with your direction, that's fine, but I want to address the minor patch first, before pulling large wide-ranging changes to the common code.

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,532 @@
+      choices:
+        - 'run'
+        - 'check'
+        - 'scan'
```
Previously, we explicitly got rid of the "scan" option. I thought the API had purged it, but it is also in the AWX collection list of changes.

beeankha committed:
Update job template module, unit tests, functional tests, fix sanity and unit test errors.

Revert tower_group.py changes

Revert changes to tower_group unit test

Rebase and update

beeankha commented:
recheck

beeankha commented:
recheck

beeankha commented:
recheck

AlanCoding commented on awx_collection/test/awx/test_team.py:
```diff
@@ -47,7 +48,10 @@ def test_modify_team(run_module, admin_user):
+    result.pop('field_changes')
     assert result == {
+        "needed_update": True,
+        "updated": False,
```
This isn't something consistent enough that we could document in the `RETURNS` string. If it won't buy us that, I don't see why we should do it.

If it's created, then it returns the `created` key, and if it's updated it returns the `updated` key. But neither of these keys are present if it's not created or updated. The proposal in https://github.com/ansible/awx/issues/6367 is that other status like `approved` would also be present. Then should `started` and `finished` be status types too?

I'd advocate that returning `state` might be best, even if it's a repeat of the input parameter. That's documentable, and not always obvious because it has a default.

john-westcott-iv replied:
I added some of this as almost "Tracking" so that you could look at the returned data and know what it did. I thought that might also help us prepare for diff mode. i.e. instead of just `changed: True` why was it actually changed. This has also helped in development of Tower credential to know why a credential is being reported as changed.

I think we should be able to set them in the returned field because they can have an optional description on them (I believe). i.e.
```
needed_update:
    description: If the item needed to be updated
    returned: when item already exists
    type: bool
    sample: True
```

AlanCoding replied:
If there's going to be churn on this, we might be better off to rewrite the (unit) test assertions so that they just don't assert on the full data - and develop something after things have stabilized. For the purposes of testing / tracking specific behaviors, that makes sense to me, I just worry if the key naming seems too general. But I also can't connect the dots from these changes to the JT module conversion. A separate PR might be nice.

john-westcott-iv replied:
The only reason they were in here was because I was hitting a condition where the JT was always returning changed = True. Instead of doing my old: `with open('/tmp/john', 'w') as f ....` I tried to get smart and add these additional fields into the module_util so that I could tell what it was trying to do internally. Incidentally, I needed these things when working on tower_credential to track down the issue with the inputs hash.

AlanCoding commented on awx_collection/plugins/modules/tower_group.py:
```diff
@@ -112,9 +112,21 @@ def main():
+    if children is not None:
+        children_ids = []
+        for item in children:
+            children_ids.append(module.resolve_name_to_id('groups', item))
```
this directly conflicts with `association_fields`, I think this would be better split out into another PR that just changes the field name, which I agree should be done.

john-westcott-iv replied:
Agreed, this PR should have just been for tower_job_template. @beeankha let me know if you want to pull this out of here or if I should?

beeankha replied:
I can do this, and put the changes into a new `tower_group.py`-specific PR.

AlanCoding replied:
resolved separately

squidboylan commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,562 @@
-        playbook=dict(required=True),
+        name=dict(required=True, type='str'),
+        new_name=dict(required=False, type='str'),
+        description=dict(required=False, type='str', default=''),
```
I'm moderately concerned about all the defaults here, mostly in the case of modifying a job_template, if we have a default and the user doesn't specify a value for that parameter we'll overwrite that parameter with the default, we do need defaults for a lot of these things, but only for new objects. I think we ran into this issue with a different module, i forget which one, maybe user

AlanCoding replied:
we've discussed this in some other places too. I'm in agreement, my stance is that `state` should have a default, and the rest can go.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -551,12 +635,12 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
 
         # If we have an on_create method and we actually changed something we can call on_create
-        if on_create is not None and self.json_output['changed']:
+        if on_create is not None and self.json_output['created']:
```
If you modify a project's URL or something like that, you still need to wait for update because it will auto-update in those cases.

beeankha committed:
Clean up unit tests

AlanCoding commented:
```python
@pytest.mark.django_db
def test_job_template_with_new_credentials(
        run_module, admin_user, project, inventory,
        machine_credential, vault_credential):
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        credentials=[machine_credential.name, vault_credential.name]
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert result.get('changed', False), result
    jt = JobTemplate.objects.get(pk=result['id'])

    assert set([machine_credential.id, vault_credential.id]) == set([
        cred.pk for cred in jt.credentials.all()])

    prior_ct = ActivityStream.objects.count()
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        credentials=[machine_credential.name, vault_credential.name]
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert not result.get('changed', True), result
    jt.refresh_from_db()
    assert result['id'] == jt.id

    assert set([machine_credential.id, vault_credential.id]) == set([
        cred.pk for cred in jt.credentials.all()])
    assert ActivityStream.objects.count() == prior_ct


@pytest.mark.django_db
def test_job_template_with_survey_spec(run_module, admin_user, project, inventory, survey_spec):
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        survey_spec=survey_spec,
        survey_enabled=True
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert result.get('changed', False), result
    jt = JobTemplate.objects.get(pk=result['id'])

    assert jt.survey_spec == survey_spec

    prior_ct = ActivityStream.objects.count()
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        survey_spec=survey_spec,
        survey_enabled=True
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert not result.get('changed', True), result
    jt.refresh_from_db()
    assert result['id'] == jt.id

    assert jt.survey_spec == survey_spec
    assert ActivityStream.objects.count() == prior_ct
```

I'd suggest adding these, replacing the existing credentials test.

I also believe that both tests should be expanded to modify the credentials and the survey spec. The credential modification scenarios are a lot, so I'd pack them all into 1 test (add, remove, stay the same). You could piggyback on these cases to these tests, or add new tests.

Right now I get:

```
awx_collection/test/awx/test_job_template.py ....F                                                                                                                            [100%]

===================================================================================== FAILURES ======================================================================================
________________________________________________________________________ test_job_template_with_survey_spec _________________________________________________________________________
Traceback (most recent call last):
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/test_job_template.py", line 149, in test_job_template_with_survey_spec
    assert jt.survey_spec == survey_spec
AssertionError: assert {} == {'description... False, ...}]}
  Right contains 3 more items:
  {'description': 'test',
   'name': 'test',
   'spec': [{'default': 'mydef',
             'index': 0,
             'question_name': 'my question?',
             'required': False,...
  
  ...Full output truncated (4 lines hidden), use '-vv' to show
============================================================================ 1 failed, 4 passed in 4.13s ============================================================================
```

AlanCoding commented on awx_collection/test/awx/test_inventory_source.py:
```diff
@@ -42,6 +42,7 @@ def test_inventory_source_create(run_module, admin_user, base_inventory, project
     assert result == {
         'id': inv_src.id,
         'name': 'foo',
+        'created': True,
```
just change the whole thing to `assert result['id'] == inv_src.id` because it is being said here that the others are not supported.

The name return value might be alright, but it's not given in all cases. No one has any idea if should be there or not.

AlanCoding commented on awx_collection/tools/vars/associations.yml:
```diff
@@ -0,0 +1,16 @@
+    - related_item: hosts
+      endpoint: hosts
+      description: "The hosts associated with this group"
+      required: false
```
I think `required` shouldn't be a thing here, and the arg spec should just leave it out.

AlanCoding commented on awx_collection/tools/vars/resolution.yml:
```diff
@@ -0,0 +1,6 @@
+  webhook_credential: credentials
+  project: projects
+  inventory: inventories
+  organization: organizations
```
I don't think we should continue to rely on name_to_id, and encoding the logic in the generator is going in the wrong directions. I would suggest that `param_to_endpoint` can just be called, and there is no need for an entire pipeline for producing this call pattern.

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,562 @@
+    if webhook_service is not None:
+        new_fields['webhook_service'] = webhook_service
+    if webhook_credential is not None:
+        new_fields['webhook_credential'] = webhook_credential_id
```
This is a lot of additional code. The old module did an outright `params.copy()`, so there was no enumeration of the field names at all, and with this PR we're losing that, and we're not getting anything for it. As a _tiny_ improvement I've listed the params in other modules to keep it slightly more brief. In this case I think the ideal solution would be to remove params that shouldn't be treated in the standard way.

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,562 @@
+        - The Tower OAuth token to use.
+      required: False
+      type: str
+      version_added: "3.7"
```
Is there any way to find out what in the parameter spec changed?

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,562 @@
-    tower_config_file: "~/tower_cli.cfg"
-    survey_enabled: yes
-    survey_spec: "{{ lookup('file', 'my_survey.json') }}"
-    custom_virtualenv: "/var/lib/awx/venv/custom-venv/"
```
the module does need an example

beeankha replied:
Oops, will add one back in...

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -12,409 +13,562 @@
+            labels_ids.append(module.resolve_name_to_id('labels', item))
+
+    # Pull in additional posts
+    additional_posts = []
```
Since they're both hitting some kind of related items, it's really mismatched that the additional posts is a list while the additional associations is a dict. It's also really hard to keep up with an implied nested schema.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -551,12 +635,12 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
                 self.modify_associations(sub_endpoint, associations[association_type])
```
You're adding a helper method to tease out "detail" or "__all__" or other stuff from the response, and to the extent that you still want to do that, branching like what we have here in `create_if_needed` should be converted to use that helper method.

AlanCoding replied:
Also, the method `create_if_needed` calls `modify_associations` while the other one calls `process_associations`. If it's going to be called in exactly 1 place, there's no reason to put it into a separate method.

beeankha committed:
Delete a tools file, update module utils, update misc unit tests

beeankha committed:
Revert rebase mistake in tower_group module

john-westcott-iv committed:
Fixing additional_posts

beeankha commented:
Closed due to https://github.com/ansible/awx/pull/6491 being merged.
